test/e2e/autoscale_test.go

```diff
@@ -200,56 +202,74 @@ func assertScaleDown(ctx *testContext) {
-	deployment, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
+func numberOfPods(ctx *testContext) (int32, error) {
+	deployment, err := ctx.clients.KubeClient.Kube.Apps().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "Failed to get deployment %q", deployment)
-	}
-	gotReplicas := deployment.Status.ReadyReplicas
-	mes := fmt.Sprintf("got %d replicas, expected between [%d, %d] replicas for deployment %s", gotReplicas, numReplicasMin, numReplicasMax, ctx.deploymentName)
-	ctx.t.Log(mes)
-	if gotReplicas < numReplicasMin || gotReplicas > numReplicasMax {
-		return errors.New(mes)
+		return 0, errors.Wrapf(err, "Failed to get deployment %q", deployment)
 	}
-	return nil
+	return deployment.Status.ReadyReplicas, nil
 }
 
-func assertAutoscaleUpToNumPods(ctx *testContext, numPods int32) {
-	// Relaxing the pod count requirement a little bit to avoid being too flaky.
-	minPods := numPods - 1
-	maxPods := numPods + 1
+func assertAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods int32, duration time.Duration, quick bool) {
+	// There are two test modes: quick, and not quick.
+	// 1) Quick mode: succeeds when the number of pods meets targetPods.
+	// 2) Not Quick (sustaining) mode: succeeds when the number of pods gets scaled to targetPods and
+	//    sustains there for the duration.
+
+	var minPods, maxPods int32
+	minPods = curPods
+	if quick {
+		// Relaxing the pod count requirement a little bit to avoid being too flaky.
+		maxPods = targetPods + 1
+	} else {
+		// Leave ample headroom because when switching to panic mode, the app is running at
+		// panicThresholdPercentage of the containerConcurrency (and panicThresholdPercentage is at least 110).
+		maxPods = targetPods*2
+	}
 
 	stopChan := make(chan struct{})
 	var grp errgroup.Group
 	grp.Go(func() error {
-		return generateTraffic(ctx, int(numPods*10), 60*time.Second, stopChan)
+		return generateTraffic(ctx, int(targetPods*containerConcurrency), duration, stopChan)
 	})
 	grp.Go(func() error {
 		// Short-circuit traffic generation once we exit from the check logic.
 		defer close(stopChan)
 
+		done := time.After(duration)
 		timer := time.Tick(2 * time.Second)
 		for {
 			select {
 			case <-timer:
-				if err := assertNumberOfPods(ctx, minPods, maxPods); err != nil {
+				got, err := numberOfPods(ctx)
+				if err != nil {
 					return err
 				}
-				if err := assertNumberOfPods(ctx, numPods, maxPods); err == nil {
-					return nil
+				mes := fmt.Sprintf("got %d replicas, expected between [%d, %d] replicas for deployment %s", got, minPods, maxPods, ctx.deploymentName)
+				ctx.t.Log(mes)
+				if got < minPods || got > maxPods {
+					return errors.New(mes)
 				}
+				if quick {
+					if got >= targetPods {
+						return nil
+					}
+				} else {
```

Comment:
`else if minPods < ...`
But probably you can just remove the `if`.
Also why not just `minPods = got`?

Reply:
`minPods = got` doesn't work, because the number of pods surges over the targetPods in panic mode, then it gets scaled down to targetPods when exiting panic mode.